oximeter/db/src/client/oxql.rs

```diff
@@ -0,0 +1,704 @@
+    ) -> Result<(QuerySummary, BTreeMap<TimeseriesKey, oxql::Timeseries>), Error>
+    {
+        // We'll create timeseries for each key on the fly. To enable computing
+        // deltas, we need to track the last measurement we've seen as well.
+        let mut measurements_by_key: BTreeMap<_, Vec<_>> = BTreeMap::new();
+        let measurements_query =
+            self.measurements_query(schema, preds, info.keys())?;
+        let mut n_measurements: usize = 0;
+        let (summary, body) =
+            self.execute_with_body(&measurements_query).await?;
+        for line in body.lines() {
+            let (key, measurement) =
+                model::parse_measurement_from_row(line, schema.datum_type);
+            measurements_by_key.entry(key).or_default().push(measurement);
+            n_measurements += 1;
+        }
+        debug!(
+            &self.log,
+            "fetched measurements for OxQL query";
+            "n_keys" => measurements_by_key.len(),
+            "n_measurements" => n_measurements,
+        );
+
+        // Remove the last measurement, returning just the keys and timeseries.
+        let mut out = BTreeMap::new();
+        for (key, measurements) in measurements_by_key.into_iter() {
+            // Constuct a new timeseries, from the target/metric info.
+            let (target, metric) = info.get(&key).unwrap();
+            let mut timeseries = oxql::Timeseries::new(
+                target
+                    .fields
+                    .iter()
+                    .chain(metric.fields.iter())
+                    .map(|field| (field.name.clone(), field.value.clone())),
+                oxql::point::DataType::try_from(schema.datum_type)?,
+                if schema.datum_type.is_cumulative() {
+                    oxql::point::MetricType::Delta
+                } else {
+                    oxql::point::MetricType::Gauge
+                },
+            )?;
+
+            // Covert its oximeter measurements into OxQL data types.
+            let points = if schema.datum_type.is_cumulative() {
+                oxql::point::Points::delta_from_cumulative(&measurements)?
+            } else {
+                oxql::point::Points::gauge_from_gauge(&measurements)?
+            };
+            timeseries.points = points;
+            debug!(
+                &self.log,
+                "inserted new OxQL timeseries";
+                "key" => key,
+                "metric_type" => ?timeseries.points.metric_type(),
+                "n_points" => timeseries.points.len(),
+            );
+            out.insert(key, timeseries);
+        }
+        Ok((summary, out))
+    }
+
+    async fn verify_measurement_query_limit<'keys>(
+        &self,
+        schema: &TimeseriesSchema,
+        preds: Option<&oxql::ast::table_ops::filter::Filter>,
+        consistent_keys: impl ExactSizeIterator<Item = &'keys TimeseriesKey>,
+    ) -> Result<(), Error> {
+        const MAX_ROWS_PER_TIMESERIES_PER_QUERY: usize = 100_000;
```

Comment:
Ugh, yeah, this was a hacky way to prevent huge queries. I think if we want configurability here, I should probably bite the bullet and implement real pagination.

If we do keep it, I'll definitely add a comment. For clarity, this is the maximum number of data points that we fetch in any one SQL query to ClickHouse. Concretely, for any `get` table operations in an OxQL query, we first check that the number of rows is no greater than this. So you can still get back more than this many points, such as if you write a nested subquery.

Reply:
See sibling comment about pagination. Going to mark this resolved, and track the related work in the other thread.